lib/galaxy/jobs/runners/kubernetes.py

```diff
@@ -157,15 +158,24 @@ def queue_job(self, job_wrapper):
         job = Job(self._pykube_api, k8s_job_obj)
         try:
             job.create()
-            job_id = job.metadata.get('name')
-            # define job attributes in the AsyncronousJobState for follow-up
-            ajs.job_id = job_id
-            # store runner information for tracking if Galaxy restarts
-            job_wrapper.set_external_id(job_id)
-        except Exception as e:
-            log.exception("Error occurred while creating kubernetes job object")
-            raise e
+        except HTTPError:
+            job_wrapper.fail("Kubernetes failed to create job.", exception=True)
+            log.exception("Kubernetes failed to create job, HTTP exception encountered")
+            ajs.runner_state = JobState.runner_states.UNKNOWN_ERROR
+            self.mark_as_failed(ajs)
+            return
+        if not job.name:
```

Comment:
I'm not sure I understand how `job.name` could be null here if the `job.create()` call succeeded. The name is a mandatory part of the k8s job spec and returned response?

Reply:
See https://github.com/galaxyproject/galaxy/issues/11714

Reply:
Yes, I was referring to #11714 in my closing review comment. In that stack trace, `job.get_job_runner_external_id()` is what's returning null right, which means there could be multiple reasons for that being null. At the same time, I can't see how `job.name` could be null without violating the API spec? https://github.com/kelproject/pykube/blob/e8a46298a592ad9037587afb707ac75b3114eff9/pykube/objects.py#L49

Reply:
I think the point is that pykube *is* violating the spec. Somehow it is getting thrown into a state the code is not handling and returns an empty name.

Regardless of the reason, it is vital to our code that job.name be populated before continuing and a check should be made.

Reply:
```suggestion
I think I remember tracing it out and finding that pykube was returning a completely unpopulated job object.
```
Right, that's what I was trying to understand. This increases the number of reasons to hasten pykube's retirement. Looking at the code though, I can't see any obvious place where PyKube itself could return an empty APIObject: https://github.com/kelproject/pykube/blob/e8a46298a592ad9037587afb707ac75b3114eff9/pykube/objects.py#L155

Could it be tied to the k8s version perhaps? Which version of k8s did you encounter this on? In any case, we'll need to work on getting ride of pykube. Ping @afgane @almahmoud

Reply:
k8s 1.8
I suppose you mean to replace it with https://github.com/kubernetes-client/python?

I am still planning to rewrite the k8s and docker runners taking a template job spec approach. I was supposed to get a new hire but it fell through and now the task has no allocated time.